### PR TITLE
Dev/singularity handling

### PR DESCRIPTION
**Pull request by chongwilliam**
Singularity handling through blending torque for type 1 and type 2 singularities.

### CODE REVIEW

File: src/RobotController.cpp [hunk 1]
```diff
@@ -59,8 +59,7 @@ Eigen::VectorXd RobotController::computeControlTorques() {
 		control_torques += task->computeTorques() - previous_tasks_disturbance;
 	}
 	previous_tasks_disturbance =
-		(MatrixXd::Identity(dof, dof) -
-		 _redundancy_completion_task->getTaskNullspace().transpose()) *
+		_redundancy_completion_task->getPreviousTasksNullspace().transpose() *
```
Comment (mikael-jorda) [inline src/RobotController.cpp:62]: I don't think this is correct. This is N_prec, but what we want here is the range space of this specific task

File: src/tasks/MotionForceTask.cpp [hunk 2]
```diff
@@ -164,6 +164,11 @@ void MotionForceTask::initialSetup() {
 		_current_position, _current_orientation, getLoopTimestep());
 	enableInternalOtgAccelerationLimited(0.3, 1.0, M_PI / 3, M_PI);
 
+	// singularity handling
+	setSingularityBounds(2e-2, 2e-3);  // default for panda 
+	MatrixXd J_posture = getConstRobotModel()->linkDependency(_link_name);
+	_singularity_handler = std::make_shared<SingularityHandler>(getConstRobotModel(), J_posture);
```
Comment (mikael-jorda) [inline src/tasks/MotionForceTask.cpp:170]: Not too important for now, but maybe use a unique_ptr instead of a shared_ptr here

File: src/tasks/SingularityHandler.h [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+ *          since the singular direction can move pi/2 either direction to be orthogonal to the force vector
+ *          since the singular direction only needs to be in-line with the force vector.
+ * @version 0.1
+ * @date 2024-01-13
+ * 
+ * @copyright Copyright (c) 2024
+ * 
+ */
+
+#pragma once
```
Comment (mikael-jorda) [inline src/tasks/SingularityHandler.h:17]: Use header guards
#ifndef ...._H_
#define ...._H_

#endif 

for consistency with the rest of the repo

File: src/tasks/SingularityHandler.h [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+    TYPE_1_SINGULARITY,
+    TYPE_2_SINGULARITY
+};
+
+class SingularityHandler {
+public:
+    SingularityHandler(std::shared_ptr<Sai2Model::Sai2Model> robot,
+                       const MatrixXd& J_posture,
+                       const double& type_1_tol = 0.95);
+    void updateTaskModel(const MatrixXd& full_jacobian,
```
Comment (mikael-jorda) [inline src/tasks/SingularityHandler.h:36]: Can we pass a shared pointer to the robot model in the constructor, and only pass the projected jacobian in this update function and do all the necessary computations inside ?

**Comment by chongwilliam** [comment]
Revised jacobian implementation subject to joint space tuning

File: src/tasks/JointTask.cpp [hunk 5]
```diff
@@ -15,15 +15,16 @@ JointTask::JointTask(std::shared_ptr<Sai2Model::Sai2Model>& robot,
 					 const std::string& task_name, const double loop_timestep)
 	: TemplateTask(robot, task_name, TaskType::JOINT_TASK, loop_timestep) {
 	// selection for full joint task
-	_joint_selection = MatrixXd::Identity(getConstRobotModel()->dof(), getConstRobotModel()->dof());
+	_joint_selection = MatrixXd::Identity(getConstRobotModel()->dof(),
```
Comment (mikael-jorda) [inline src/tasks/JointTask.cpp:18]: It looks like all the changes in JointTask are not from you, but you need to merge/rebase on master

File: src/tasks/MotionForceTask.cpp [hunk 6]
```diff
@@ -96,6 +96,7 @@ MotionForceTask::MotionForceTask(
 
 void MotionForceTask::initialSetup() {
 	setDynamicDecouplingType(BOUNDED_INERTIA_ESTIMATES);
+	// setDynamicDecouplingType(FULL_DYNAMIC_DECOUPLING);  // debug
```
Comment (mikael-jorda) [inline src/tasks/MotionForceTask.cpp:99]: remove

File: src/tasks/MotionForceTask.cpp [hunk 7]
```diff
@@ -164,7 +165,12 @@ void MotionForceTask::initialSetup() {
 		_current_position, _current_orientation, getLoopTimestep());
 	enableInternalOtgAccelerationLimited(0.3, 1.0, M_PI / 3, M_PI);
 
-	reInitializeTask();
+	// singularity handling
+	MatrixXd J_posture = getConstRobotModel()->linkDependency(_link_name);
```
Comment (mikael-jorda) [inline src/tasks/MotionForceTask.cpp:169]: I still don't get why we need this jacobian and why we cannot just take the "normal" jacobian

File: src/tasks/MotionForceTask.cpp [hunk 8]
```diff
@@ -221,39 +227,55 @@ void MotionForceTask::updateTaskModel(const MatrixXd& N_prec) {
...
-		_current_task_range.block(3, _pos_range, 3, _ori_range) = range_ori;
-	}
-
-	Sai2Model::OpSpaceMatrices op_space_matrices =
-		getConstRobotModel()->operationalSpaceMatrices(
-			_current_task_range.transpose() * _projected_jacobian);
-	_Lambda = op_space_matrices.Lambda;
-	_Jbar = op_space_matrices.Jbar;
-	_N = op_space_matrices.N;
+	// /*
```
Comment (mikael-jorda) [inline src/tasks/MotionForceTask.cpp:230]: remove the old implementation, don't let it commented

File: src/tasks/MotionForceTask.cpp [hunk 9]
```diff
@@ -221,39 +227,55 @@ void MotionForceTask::updateTaskModel(const MatrixXd& N_prec) {
...
+	// 	getConstRobotModel()->operationalSpaceMatrices(
+	// 		_current_task_range.transpose() * _projected_jacobian);
+	// _Lambda = op_space_matrices.Lambda;
+	// _Jbar = op_space_matrices.Jbar;
+	// _N = op_space_matrices.N;
+
+	/*
+		Singularity handler implementation 
+	*/
+	SingularityOpSpaceMatrices singularity_op_space_matrices = _singularity_handler->updateTaskModel(_projected_jacobian, _N_prec);
```
Comment (mikael-jorda) [inline src/tasks/MotionForceTask.cpp:266]: I suggest keeping those opspace matrices as member variables inside the singularity handler (which you already have), and having getters for those you need.

In addition, I suggest letting the singularity handler know about the dynamic decoupling type, and do the related computations inside the updateTaskModel of the singularity handler directly. You can for example pass the decoupling type to the updateTask model of the singularity handler.

Finally, you can have the modified lambda computed directly inside the new class (keep them as separate variables from the non modified lambda), and you don't need the setLambda function

And if you do that, there is no need to for all those following lines, Just the _N matrix update from the singularity handler, and the updateTaskModel of the singularuty handler class can return void

File: src/tasks/MotionForceTask.cpp [hunk 10]
```diff
@@ -221,39 +227,55 @@ void MotionForceTask::updateTaskModel(const MatrixXd& N_prec) {
...
+	_projected_jacobian_ns = singularity_op_space_matrices.projected_jacobian_ns;
+	_Lambda_ns = singularity_op_space_matrices.Lambda_ns;
+	_task_range_ns = singularity_op_space_matrices.task_range_ns;
+	_projected_jacobian_s = singularity_op_space_matrices.projected_jacobian_s;
+	_Lambda_s = singularity_op_space_matrices.Lambda_s;
+	_task_range_s = singularity_op_space_matrices.task_range_s;
 
 	switch (_dynamic_decoupling_type) {
```
Comment (mikael-jorda) [inline src/tasks/MotionForceTask.cpp:274]: The more I think about it, the more I think we should remove the partial dynamic decoupling. Please do it if you move the decoupling computations inside the new class as I suggested before

File: src/tasks/MotionForceTask.h [hunk 11]
```diff
@@ -519,6 +527,36 @@ class MotionForceTask : public TemplateTask {
...
+	 * @brief	Change the bounds for singularity cut-off
+	 * 
+	 * @param e_max 
+	 * @param e_min 
+	 */
+	void setSingularityBounds(const double& s_min, const double& s_max) {
+		_singularity_handler->setSingularityBounds(s_min, s_max);
+	}
+
+	MatrixXd getProjectedJacobian() {
```
Comment (mikael-jorda) [inline src/tasks/MotionForceTask.h:540]: why do we need this getter?

File: src/tasks/MotionForceTask.h [hunk 12]
```diff
@@ -519,6 +527,36 @@ class MotionForceTask : public TemplateTask {
...
+		return _projected_jacobian_ns;
+	}
+
+	VectorXd getUnitControlForces() {
+		return _unit_mass_force;
+	}
+
+	VectorXd getControlForces() {
```
Comment (mikael-jorda) [inline src/tasks/MotionForceTask.h:548]: why do we need this getter ?

File: src/tasks/MotionForceTask.h [hunk 13]
```diff
@@ -519,6 +527,36 @@ class MotionForceTask : public TemplateTask {
...
+		return _Lambda_ns_modified * _task_range_ns.transpose() * _unit_mass_force;
+	}
+
+	MatrixXd getLambdaMatrix() {
```
Comment (mikael-jorda) [inline src/tasks/MotionForceTask.h:552]: This seems incorrect. the returned Lambda matrix won't be symmetric, and maybe not even square. Also, do we need this getter ?

File: src/tasks/MotionForceTask.h [hunk 14]
```diff
@@ -519,6 +527,36 @@ class MotionForceTask : public TemplateTask {
...
+		return _Lambda_ns_modified * _task_range_ns.transpose();
+	}
+
+	const std::unique_ptr<SingularityHandler>& getConstSingularityModel() const {
```
Comment (mikael-jorda) [inline src/tasks/MotionForceTask.h:556]: I am not sure we want this getter either, why would the user need acces to the internal singularity handler ?

File: src/tasks/MotionForceTask.h [hunk 15]
```diff
@@ -527,6 +565,12 @@ class MotionForceTask : public TemplateTask {
 	 */
 	void initialSetup();
 
+	// singularity handling
+	std::unique_ptr<SingularityHandler> _singularity_handler;
+	MatrixXd _Lambda_s, _Lambda_ns, _Lambda_s_modified, _Lambda_ns_modified;
```
Comment (mikael-jorda) [inline src/tasks/MotionForceTask.h:570]: As mentioned before, we probably don't need those as they are already in the singularity handler class

File: src/tasks/SingularityHandler.cpp [hunk 16]
```diff
@@ -0,0 +1,249 @@
...
+
+    // classify singularities 
+    classifySingularity(_task_range_s);
+
+    // task updates 
+    if (_sing_type == NO_SINGULARITY) {
+        _N = _N_ns;  // _N = N_ns (rank(N_ns) = task_rank)
+    } else if (_sing_type != NO_SINGULARITY) {
+        // update posture task 
+        _posture_projected_jacobian = _J_posture * _N_ns * N_prec;
```
Comment (mikael-jorda) [inline src/tasks/SingularityHandler.cpp:146]: Why not using _projected_jacobian_s and avoid to have to compute yet another svd

File: src/tasks/SingularityHandler.cpp [hunk 17]
```diff
@@ -0,0 +1,249 @@
...
+
+}
+
+VectorXd SingularityHandler::computeTorques(const VectorXd& unit_mass_force, const VectorXd& force_related_terms)
+{
+    VectorXd joint_strategy_torques = VectorXd::Zero(_robot->dof());
+    VectorXd unit_torques = VectorXd::Zero(_robot->dof());
+
+    // debug
+    if (_sing_type != NO_SINGULARITY) {
```
Comment (mikael-jorda) [inline src/tasks/SingularityHandler.cpp:208]: remove those print statements before pushing

File: src/tasks/SingularityHandler.cpp [hunk 18]
```diff
@@ -0,0 +1,249 @@
...
+        Sai2Model::OpSpaceMatrices s_matrices =
+            _robot->operationalSpaceMatrices(_projected_jacobian_s);
+        _Lambda_s = s_matrices.Lambda;
+        _Jbar_s = s_matrices.Jbar;
+        _N_s = s_matrices.N;
+    } else {
+        // includes the partial task projection handling 
+        for (int i = 1; i < _task_rank; ++i) {
+            double condition_number = J_svd.s(i) / J_svd.s(0);
+            _alpha = std::clamp((condition_number - _s_min) / (_s_max - _s_min), 0., 1.);
```
Comment (mikael-jorda) [inline src/tasks/SingularityHandler.cpp:95]: compute alpha within the if statement for clearer code

**Review by mikael-jorda — COMMENTED** [review]
It looks nice overall
I have some comments about the interface and how to clean the code a little
I did not review the example yet

File: .gitignore [hunk 19]
```diff
@@ -32,7 +32,11 @@
 *.app
 
 # folders
+**.DS_Store
 **/build/
+.vscode/
+
+# Unrelated files
```
Comment (mikael-jorda) [inline .gitignore:39]: put those in .git/info/exclude file

File: src/tasks/JointTask.cpp [hunk 20]
```diff
@@ -381,4 +381,15 @@ void JointTask::enableVelocitySaturation(const VectorXd& saturation_velocity) {
 	_saturation_velocity = saturation_velocity;
 }
 
-} /* namespace Sai2Primitives */
+bool JointTask::goalPositionReached(const double& tol) {
```
Comment (mikael-jorda) [inline src/tasks/JointTask.cpp:384]: make a separate PR for those changes

File: src/tasks/JointTask.h [hunk 21]
```diff
@@ -344,6 +344,18 @@ class JointTask : public TemplateTask {
 		_dynamic_decoupling_type = type;
 	}
 
+	/**
```
Comment (mikael-jorda) [inline src/tasks/JointTask.h:347]: make a separate PR for those changes

File: src/tasks/MotionForceTask.h [hunk 22]
```diff
@@ -606,6 +586,60 @@ class MotionForceTask : public TemplateTask {
...
+	void setType1Behavior(const bool flag) {
+		_singularity_handler->setType1Behavior(flag);
+	}
+
+	/**
+	 * @brief Enables or disables joint handling in singularity handler  
+	 * 
+	 * @param flag true to enforce joint handling behavior 
+	 */
+	void enableJointHandling(const bool flag) {
```
Comment (mikael-jorda) [inline src/tasks/MotionForceTask.h:625]: call this enableSingularityHandling

File: src/tasks/MotionForceTask.h [hunk 23]
```diff
@@ -606,6 +586,60 @@ class MotionForceTask : public TemplateTask {
...
+		_singularity_handler->setDynamicDecouplingType(type);
+	}
+
+    /**
+     * @brief Enforces type 1 handling behavior if set to true, otherwise handle 
+     * type 1 or type 2 as usual
+     * 
+     * @param flag true to enforce type 1 handling behavior 
+     */
+	void setType1Behavior(const bool flag) {
```
Comment (mikael-jorda) [inline src/tasks/MotionForceTask.h:607]: handleAllSingularitiesAsType1 (and update comments)

File: src/tasks/SingularityHandler.cpp [hunk 24]
```diff
@@ -0,0 +1,346 @@
...
+            for (int i = 0; i < _joint_task_range_s.rows(); ++i) {
+                if (_joint_task_range_s(i, 0) != 0) {
+                    if (std::abs(_robot->q()(i) - _q_upper(i)) < 5 * M_PI / 180) {
+                        _type_2_direction(i) = - 1;
+                    } else if (std::abs(_robot->q()(i) - _q_lower(i)) < 5 * M_PI / 180) {
+                        _type_2_direction(i) = 1;
+                    } 
+                }
+            }
+            double fTd = ((unit_mass_force.normalized() + force_related_terms.normalized()).normalized()).dot(_task_range_s.col(0));
```
Comment (mikael-jorda) [inline src/tasks/SingularityHandler.cpp:327]: probably more correct to sum the unit mass force and the force related term, and only normalize their sum

File: src/tasks/SingularityHandler.h [hunk 25]
```diff
@@ -0,0 +1,214 @@
...
+     */
+    void updateTaskModel(const MatrixXd& projected_jacobian, const MatrixXd& N_prec);
+
+    /**
+     * @brief Classifies the singularity based on a joint perturbation in the singular joint space 
+     * 
+     * @param singular_task_range Singular task range corresponding to the columns of U from SVD
+     * @param singular_joint_task_range Singular task range corresponding to the columns of V from SVD
+     */
+    void classifySingularity(const MatrixXd& singular_task_range,
```
Comment (mikael-jorda) [inline src/tasks/SingularityHandler.h:82]: this should be a private function

File: src/helper_modules/Sai2PrimitivesCommonDefinitions.h [hunk 26]
```diff
@@ -2,6 +2,7 @@
 #define SAI2_PRIMITIVES_COMMON_DEFINITIONS_H_
 
 #include <Eigen/Dense>
+#include <vector>
```
Comment (mikael-jorda) [inline src/helper_modules/Sai2PrimitivesCommonDefinitions.h:5]: put this in another PR

Comment (chongwilliam) [reply src/helper_modules/Sai2PrimitivesCommonDefinitions.h:5]: done

File: src/tasks/MotionForceTask.h [hunk 27]
```diff
@@ -664,6 +654,69 @@ class MotionForceTask : public TemplateTask {
...
+	void setType1Posture(const VectorXd& q_des) {
+		_singularity_handler->setType1Posture(q_des);
+	}
+
+	/**
+	 * @brief Enables or disables singularity handling in singularity handler  
+	 * 
+	 * @param flag true to enable singularity handling, if false the behavior is just task truncation 
+	 */
+	void enableSingularityHandling(const bool flag) {
```
Comment (mikael-jorda) [inline src/tasks/MotionForceTask.h:693]: make 2 functions
enableSingularityHandling()
disableSingularityHandling()
that take no argument. I think it is cleaner and mimics the velocity saturation and otg enable/disable

File: src/tasks/MotionForceTask.h [hunk 28]
```diff
@@ -664,6 +654,69 @@ class MotionForceTask : public TemplateTask {
...
+
+    /**
+     * @brief Set the singularity bounds for torque blending based on the inverse of the condition number
+     * The linear blending coefficient \alpha is computed as \alpha = (s - _s_min) / (_s_max - _s_min),
+     * and is clamped between 0 and 1.
+     * 
+     * @param s_min lower bound
+     * @param s_max upper bound 
+     */
+	void setSingularityBounds(const double& s_min, const double& s_max) {
```
Comment (mikael-jorda) [inline src/tasks/MotionForceTask.h:705]: rename to setSingularityHandlingBounds

**Review by mikael-jorda — APPROVED** [review]
